energymeters/energymeter.cpp

```diff
@@ -0,0 +1,206 @@
+* Alternatively, this project may be redistributed and/or modified under the
+* terms of the GNU Lesser General Public License as published by the Free
+* Software Foundation; version 3. This project is distributed in the hope that
+* it will be useful, but WITHOUT ANY WARRANTY; without even the implied
+* warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public License
+* along with this project. If not, see <https://www.gnu.org/licenses/>.
+*
+* For any further details and any questions please contact us under
+* contact@nymea.io or see our FAQ/Licensing Information on
+* https://nymea.io/license/faq
+*
+* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+#include "energymeter.h"
+#include "hardware/modbus/modbusrtureply.h"
+
+#include "extern-plugininfo.h"
+
+EnergyMeter::EnergyMeter(ModbusRtuMaster *modbusMaster, int slaveAddress, const QHash<ModbusRegisterType, ModbusRegisterDescriptor> &modbusRegisters, QObject *parent) :
+    QObject(parent),
+    m_modbusRtuMaster(modbusMaster),
+    m_slaveAddress(slaveAddress),
+    m_modbusRegisters(modbusRegisters)
+{
+
+}
+
+QUuid EnergyMeter::modbusRtuMasterUuid()
+{
+    return m_modbusRtuMaster->modbusUuid();
+}
+
+bool EnergyMeter::connected()
+{
+    return m_connected;
+}
+
+bool EnergyMeter::getVoltageL1()
+{
+    return getRegister(ModbusRegisterType::VoltageL1);
+}
+
+bool EnergyMeter::getVoltageL2()
+{
+    return getRegister(ModbusRegisterType::VoltageL2);
+}
+
+bool EnergyMeter::getVoltageL3()
+{
+    return getRegister(ModbusRegisterType::VoltageL3);
+}
+
+bool EnergyMeter::getCurrentL1()
+{
+    return getRegister(ModbusRegisterType::CurrentL1);
+}
+
+bool EnergyMeter::getCurrentL2()
+{
+    return getRegister(ModbusRegisterType::CurrentL2);
+}
+
+bool EnergyMeter::getCurrentL3()
+{
+    return getRegister(ModbusRegisterType::CurrentL3);
+}
+
+bool EnergyMeter::getFrequency()
+{
+    return getRegister(ModbusRegisterType::Frequency);
+}
+
+bool EnergyMeter::getPowerFactor()
+{
+    return getRegister(ModbusRegisterType::PowerFactor);
+}
+
+bool EnergyMeter::getActivePower()
+{
+    return getRegister(ModbusRegisterType::ActivePower);
+}
+
+bool EnergyMeter::getEnergyProduced()
+{
+    return getRegister(ModbusRegisterType::EnergyProduced);
+}
+
+bool EnergyMeter::getEnergyConsumed()
+{
+    return getRegister(ModbusRegisterType::EnergyConsumed);
+}
+
+bool EnergyMeter::getRegister(ModbusRegisterType type)
+{
+    if (!m_modbusRegisters.contains(type))
+        return false;
+
+    ModbusRegisterDescriptor descriptor = m_modbusRegisters.value(type);
+
+    ModbusRtuReply *reply = nullptr;
+    if (descriptor.functionCode() == 3){
+        reply = m_modbusRtuMaster->readHoldingRegister(m_slaveAddress, descriptor.address(), descriptor.length());
+    } else if (descriptor.functionCode() == 4){
```

Comment:
probably just `else`, not `else if`, otherwise you could run into a situation where `reply` is still `nullptr`.